pkg/auth/jwt_auth_cache.go

```diff
@@ -0,0 +1,78 @@
+package auth
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/grpc-ecosystem/go-grpc-middleware/auth"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type jwtAuthCache struct {
+	headerScheme string
+	jwtSecret    []byte
+	tokenCache   map[string]time.Time
+	cacheOrder   []string
+	cacheSize    int
+}
+
+// NewJwtAuthCache creates an AuthCache that implements bearer style
+// JWT based authorization validation.
+func NewJwtAuthCache(secretFile string) (AuthCache, error) {
+	data, err := ioutil.ReadFile(secretFile)
+	if err != nil {
+		return nil, err
+	}
+	return &jwtAuthCache{
+		headerScheme: "bearer",
+		jwtSecret:    bytes.Trim(data, "\n"),
+		tokenCache:   make(map[string]time.Time),
+		cacheOrder:   make([]string, 0),
+		cacheSize:    100,
+	}, nil
+}
+
+func BuildJWTAuthFunction(authCache AuthCache) func(ctx context.Context) (context.Context, error) {
```

Comment:
I think we could do without this helper function. The only difference between `authCache.ValidateCredentials` and the type of the function returned by this function is that this one returns an additional context.

What if `ValidateCredentials()` had the same prototype? In that case you could use `authCache.ValidateCredentials` directly, instead of `BuildJWTAuthFunction(authCache)`. This would also allow you to eliminate the `AuthCache` interface.